x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeSinkHandler.java

```diff
@@ -150,6 +156,10 @@ private void notifyListeners() {
         }
     }
 
+    public void setSource(ExchangeSourceHandler sub) {
```

Comment:
I think we should avoid directly coupling the source with the sink. Could you explain why this is necessary?

Reply:
To be able to cancel all three source-sink pipelines on the remote. Without this, only the first (topmost) sink get stopped, but the pipelines under it continue to run.

Reply:
I think we should already be propagating the cancellation or completion to upstream drivers. There might be a bug where they are not wired properly.

Reply:
It didn't propagate properly when I tried it, that's why I created this workaround. I could add some details of what exactly I am seeing if needed.

Reply:
I think I understand the issue and will be working on the fix.

Reply:
I opened: https://github.com/elastic/elasticsearch/pull/120020 for this. We can make a smaller change to link the exchanges: https://github.com/elastic/elasticsearch/pull/120020/commits/b06f402c6a604209a5ecff4d942d034d3f80b927, but it will only solve cases where the Driver is put to sleep(because no pages are available in the Source) and it won't address cases where the thread pool is busy running other tasks ahead of the Driver.